File: subprojects/frida-python/examples/portal_client.py
import json
import sys

from frida_tools.application import Reactor

import frida


class Application:
    def __init__(self, nick):
        self._reactor = Reactor(run_until_return=self._process_input)

        token = {"nick": nick, "secret": "knock-knock"}
        self._device = frida.get_device_manager().add_remote_device("::1", token=json.dumps(token))

        self._bus = self._device.bus
        self._bus.on("message", lambda *args: self._reactor.schedule(lambda: self._on_bus_message(*args)))

        self._channel = None
        self._prompt = "> "

    def run(self):
        self._reactor.schedule(self._start)
        self._reactor.run()

    def _start(self):
        self._bus.attach()

    def _process_input(self, reactor):
        while True:
            sys.stdout.write("\r")
            try:
                text = input(self._prompt).strip()
            except:
                self._reactor.cancel_io()
                return
            sys.stdout.write("\033[1A\033[K")
            sys.stdout.flush()

            if len(text) == 0:
                self._print("Processes:", self._device.enumerate_processes())
                continue

            if text.startswith("/join "):
                if self._channel is not None:
                    self._bus.post({"type": "part", "channel": self._channel})
                channel = text[6:]
                self._channel = channel
                self._prompt = f"{channel} > "
                self._bus.post({"type": "join", "channel": channel})
                continue

            if text.startswith("/announce "):
                self._bus.post({"type": "announce", "text": text[10:]})
                continue

            if self._channel is not None:
                self._bus.post({"channel": self._channel, "type": "say", "text": text})
            else:
                self._print("*** Need to /join a channel first")

    def _on_bus_message(self, message, data):
        mtype = message["type"]
        if mtype == "welcome":
            self._print("*** Welcome! Available channels:", repr(message["channels"]))
        elif mtype == "membership":
            self._print("*** Joined", message["channel"])
            self._print(
                "- Members:\n\t"
                + "\n\t".join([f"{m['nick']} (connected from {m['address']})" for m in message["members"]])
            )
            for item in message["history"]:
                self._print(f"<{item['sender']}> {item['text']}")
        elif mtype == "join":
            user = message["user"]
            self._print(f"👋 {user['nick']} ({user['address']}) joined {message['channel']}")
        elif mtype == "part":
            user = message["user"]
            self._print(f"🚪 {user['nick']} ({user['address']}) left {message['channel']}")
        elif mtype == "chat":
            self._print(f"<{message['sender']}> {message['text']}")
        elif mtype == "announce":
            self._print(f"📣 <{message['sender']}> {message['text']}")
        else:
            self._print("Unhandled message:", message)

    def _print(self, *words):
        print("\r\033[K" + " ".join([str(word) for word in words]))
        sys.stdout.write(self._prompt)
        sys.stdout.flush()


if __name__ == "__main__":
    nick = sys.argv[1]
    app = Application(nick)
    app.run()

File: subprojects/frida-node/examples/portal_client.ts
import frida from "frida";
import readline from "readline";
import { inspect } from "util";

class Application {
    #nick: string;
    #channel: string | null = null;
    #prompt = "> ";

    #device: frida.Device | null = null;
    #bus: frida.Bus | null = null;
    #input: readline.Interface | null = null;

    constructor(nick: string) {
        this.#nick = nick;
    }

    async run() {
        const token = {
            nick: this.#nick,
            secret: "knock-knock"
        };
        this.#device = await frida.getDeviceManager().addRemoteDevice("::1", {
            token: JSON.stringify(token)
        });

        const bus = this.#device.bus;
        this.#bus = bus;
        bus.detached.connect(this.#onBusDetached);
        bus.message.connect(this.#onBusMessage);
        await bus.attach();

        const input = readline.createInterface({
            input: process.stdin,
            output: process.stdout,
            terminal: true
        });
        this.#input = input;
        input.on("close", this.#onStdinClosed);
        input.on("line", this.#onStdinCommand);

        this.#showPrompt();
    }

    #quit() {
        const bus = this.#bus;
        if (bus !== null) {
            bus.detached.disconnect(this.#onBusDetached);
            bus.message.disconnect(this.#onBusMessage);
            this.#bus = null;
        }

        const input = this.#input;
        if (input !== null) {
            input.close();
            this.#input = null;
        }
    }

    #onStdinClosed = () => {
        this.#quit();
    };

    #onStdinCommand = async (command: string) => {
        const bus = this.#bus!;

        try {
            process.stdout.write("\x1B[1A\x1B[K");

            if (command.length === 0) {
                this.#print("Processes:", await this.#device!.enumerateProcesses());
                return;
            }

            if (command.startsWith("/join ")) {
                if (this.#channel !== null) {
                    bus.post({
                        type: "part",
                        channel: this.#channel
                    });
                }

                const channel = command.substr(6);
                this.#channel = channel;

                this.#prompt = `${channel} > `;

                bus.post({
                    type: "join",
                    channel: channel
                });

                return;
            }

            if (command.startsWith("/announce ")) {
                bus.post({
                    type: "announce",
                    text: command.substr(10)
                });

                return;
            }

            if (this.#channel !== null) {
                bus.post({
                    channel: this.#channel,
                    type: "say",
                    text: command
                });
            } else {
                this.#print("*** Need to /join a channel first");
            }
        } catch (e) {
            this.#print(e);
        } finally {
            this.#showPrompt();
        }
    };

    #onBusDetached = () => {
        this.#quit();
    };

    #onBusMessage = (message: PortalServerMessage, data: Buffer | null) => {
        switch (message.type) {
            case "welcome": {
                this.#print("*** Welcome! Available channels:", message.channels);

                break;
            }
            case "membership": {
                this.#print("*** Joined", message.channel);

                const membersSummary = message.members.map(m => `${m.nick} (connected from ${m.address})`).join("\n\t");
                this.#print("- Members:\n\t" + membersSummary);

                for (const item of message.history)
                    this.#print(`<${item.sender}> ${item.text}`);

                break;
            }
            case "join": {
                const { user } = message;
                this.#print(`👋 ${user.nick} (${user.address}) joined ${message.channel}`);

                break;
            }
            case "part": {
                const { user } = message;
                this.#print(`🚪 ${user.nick} (${user.address}) left ${message.channel}`);

                break;
            }
            case "chat": {
                this.#print(`<${message.sender}> ${message.text}`);

                break;
            }
            case "announce": {
                this.#print(`📣 <${message.sender}> ${message.text}`);

                break;
            }
            default: {
                this.#print("Unhandled message:", message);

                break;
            }
        }
    };

    #showPrompt() {
        process.stdout.write("\r\x1B[K" + this.#prompt);
    }

    #print(...words: any[]) {
        const text = words.map(w => (typeof w === "string") ? w : inspect(w, { colors: true })).join(" ");
        process.stdout.write(`\r\x1B[K${text}\n${this.#prompt}`);
    }
}

const nick = process.argv[2];
const app = new Application(nick);
app.run()
    .catch(e => {
        console.error(e);
    });
